src/wp-admin/includes/class-wp-plugin-dependencies.php

```diff
@@ -0,0 +1,634 @@
+			$this->dependencies_parents[ $dependency['file'] ] = array();
+		}
+		$this->dependencies_parents[ $dependency['file'] ][] = $plugin;
+
+		return true;
+	}
+
+	/**
+	 * Add notices.
+	 *
+	 * @since 5.9.0
+	 * @access public
+	 *
+	 * @return void
+	 */
+	public function admin_notices() {
+		// Early return if there are no notices to display.
+		if ( empty( $this->notices ) ) {
+			return;
+		}
+
+		foreach ( $this->notices as $notice ) {
+			echo '<div class="notice notice-warning plugin-dependencies"><p>' . wp_kses_post( $notice['content'] ) . '</p></div>';
+		}
+	}
+
+	/**
+	 * Cancel plugin's activation request.
+	 *
+	 * @since 5.9.0
+	 * @access public
+	 *
+	 * @return void
+	 */
+	public function cancel_activation_request() {
+		if ( empty( $_GET['action'] ) || 'cancel-activate' !== $_GET['action'] || empty( $_GET['plugin'] ) ) {
+			return;
+		}
+		$file = sanitize_text_field( wp_unslash( $_GET['plugin'] ) );
+		check_admin_referer( 'cancel-activate-plugin_' . $file );
+
+		$this->remove_plugin_from_queue( $file );
+	}
+
+	/**
+	 * Filters the action links displayed for each plugin in the Plugins list table.
+	 *
+	 * @since 5.9.0
+	 * @access public
+	 *
+	 * @param string[] $actions     An array of plugin action links.
+	 * @param string   $plugin_file Path to the plugin file relative to the plugins directory.
+	 * @param array    $plugin_data An array of plugin data. See `get_plugin_data()`.
+	 *
+	 * @return string[]
+	 */
+	public function plugin_action_links( $actions, $plugin_file, $plugin_data ) {
+
+		$pending_activation     = in_array( $plugin_file, $this->get_plugins_to_activate(), true );
+		$has_dependencies       = ! empty( $this->get_plugin_dependencies( $plugin_file ) );
+		$in_circular_dependency = $this->in_circular_dependency( $plugin_file );
+
+		// Remove deactivation link from dependencies.
+		if ( ! empty( $this->dependencies_parents[ $plugin_file ] ) && ! $in_circular_dependency ) {
+			unset( $actions['deactivate'] );
+		}
+
+		// On plugins with unmet dependencies that the user has already requested for the plugin's activation,
+		// removes the activation link from its actions and add a "Cancel pending activation" link in its place.
+		if ( $pending_activation && $has_dependencies ) {
+			unset( $actions['activate'] );
+			if ( current_user_can( 'activate_plugin', $plugin_file ) ) {
+				$cancel_activation = sprintf(
+					'<a href="%s" class="cancel-activate unmet-dependencies" aria-label="%s">%s</a>',
+					wp_nonce_url( 'plugins.php?action=cancel-activate&amp;plugin=' . rawurlencode( $plugin_file ), 'cancel-activate-plugin_' . $plugin_file ),
+					/* translators: %s: Plugin name. */
+					esc_attr( sprintf( _x( 'Cancel activation of %s', 'plugin' ), get_plugin_data( WP_PLUGIN_DIR . '/' . $plugin_file )['Name'] ) ),
+					__( 'Cancel activation request' )
+				);
+
+				$actions = array_merge( array( 'cancel-activation' => $cancel_activation ), $actions );
```

Comment:
Could this be written as:

```php
$actions['cancel-activation'] = $cancel_activation;
```

?

Should yield the same result, but is more performant than using `array_merge`.

Reply:
The difference is that if we do `$actions['cancel-activation'] = $cancel_activation;`, the cancel-activation action is moved to the end of the array and is therefore displayed _after_ the "delete" action (and any other actions that may have been added from other filters. I agree that `array_merge_ is sub-optimal, but it yields the correct result in the correct order. There are other ways to do it, but this is the simplest solution I could see. :thinking: 
I added a comment above the line to clarify it.